Lombiq.HelpfulLibraries.OrchardCore/TagHelpers/EditorFieldSetTagHelper.cs

```diff
@@ -0,0 +1,121 @@
+using System;
+using System.Threading.Tasks;
+
+namespace Lombiq.HelpfulLibraries.OrchardCore.TagHelpers;
+
+[HtmlTargetElement("fieldset", Attributes = "asp-for")]
+public class EditorFieldSetTagHelper : TagHelper
+{
+    private readonly IHtmlGenerator _htmlGenerator;
+
+    [HtmlAttributeNotBound]
+    [ViewContext]
+    public ViewContext ViewContext { get; set; }
+
+    [HtmlAttributeName("asp-for")]
+    public ModelExpression For { get; set; }
+
+    [HtmlAttributeName("label")]
+    public LocalizedHtmlString Label { get; set; }
+
+    [HtmlAttributeName("hint")]
+    public LocalizedHtmlString Hint { get; set; }
+
+    [HtmlAttributeName("type")]
+    public string InputType { get; set; } = "text";
+
+    [HtmlAttributeName("required")]
+    public bool IsRequired { get; set; }
+
+    public EditorFieldSetTagHelper(IHtmlGenerator htmlGenerator) =>
+        _htmlGenerator = htmlGenerator;
+
+    public override Task ProcessAsync(TagHelperContext context, TagHelperOutput output)
+    {
+        if (output.Attributes.TryGetAttribute("class", out var classAttribute))
+        {
+            var newValue = $"{classAttribute.Value} form-group";
+            output.Attributes.Remove(classAttribute);
+            output.Attributes.Add("class", newValue);
+        }
+        else
+        {
+            output.Attributes.Add("class", "form-group mb-3 col-xl-6");
+        }
+
+        var isCheckbox = InputType.EqualsOrdinalIgnoreCase("checkbox");
+
+        if (isCheckbox) output.Content.AppendHtml("<div class=\"custom-control custom-checkbox\">");
+
+        var label = _htmlGenerator.GenerateLabel(
+            ViewContext,
+            For.ModelExplorer,
+            For.Name,
+            Label.Html().Trim() + (IsRequired ? " *" : string.Empty),
+            htmlAttributes: null);
+
+        if (isCheckbox)
+        {
+            var input = _htmlGenerator.GenerateCheckBox(
+                ViewContext,
+                For.ModelExplorer,
+                For.Name,
+                For.Model switch
+                {
+                    null => null,
+                    bool value => value,
+                    _ => bool.TryParse(For.Model.ToString(), out var parsedValue) ? parsedValue : null,
+                },
+                new { @class = "custom-control-input" });
+
+            label.Attributes["class"] = "custom-control-label";
+
+            AppendContent(output, input);
+            output.Content.AppendHtml("&nbsp;");
+            AppendContent(output, label);
+        }
+        else
+        {
+            var input = _htmlGenerator.GenerateTextBox(
+                ViewContext,
+                For.ModelExplorer,
+                For.Name,
+                For.Model,
+                For.ModelExplorer.Metadata.EditFormatString,
+                new { @class = "form-control", type = InputType });
+
+            if (IsRequired)
+            {
+                input.Attributes.Add("required", "required");
+            }
```

Comment:
This is valid for checkboxes, too, so move it below the branch.

Adding `required` to the checkbox:

![image](https://github.com/Lombiq/Helpful-Libraries/assets/164360/1559de0e-a072-400d-b076-6c708d43ab72)

... generates this validation message when hitting `Save`:

![image](https://github.com/Lombiq/Helpful-Libraries/assets/164360/0882d497-1166-4ad3-b7e9-22951a82aa0a)

Reply:
Now, looking at this, I'm wondering why the message from the `data-val-required` attribute is not showing anywhere 🤔 Looks like an error message for unobtrusive frontend validation which is not triggered. That doesn't make sense: it should either fire, or not be there at all, don't you think?

Reply:
Ok added, but required on a checkbox sounds incredibly cursed. I don't know anything about how unobtrusive validation works, so I can't really comment on that. Perhaps they have a special case for checkbox, because the actual text in the data-val-required seems pretty nonsensical to me.

Reply:
The validation message is ASP.NET MVC's default message for a missing but required viewmodel property, and the name is from here:

![image](https://github.com/Lombiq/Helpful-Libraries/assets/164360/5f4fb613-07c0-4b80-9d33-514ed4e1d9f7)

I guess it's there because unobtrusive validation is somehow enabled by default. [Looks like](https://stackoverflow.com/q/55582142/177710) it can be disabled using:

```c#
public void ConfigureServices(IServiceCollection services)
{
    services
        .AddMvc()
        .AddViewOptions(options =>
        {
            options.HtmlHelperOptions.ClientValidationEnabled = false;
        });
}
```
Not that we would want to.